### PR TITLE
Save contact data

### PR DESCRIPTION
**Pull request by gsvr**

### CODE REVIEW

**Comment by gsvr** [comment]
@imsickofmaps Please review

We should move the extraction of the DOB to utils after utils issue is merged.

File: go-app-personal.js [hunk 1]
```diff
@@ -187,13 +210,43 @@ go.app = function() {
...
                 },
 
-                next: function() {
-                    return {
-                        name: 'states:end_success',
-                        creator_opts: {
-                            retry: opts.retry
-                        }
-                    };
+                next: function(content) {
```
Comment (imsickofmaps) [inline go-app-personal.js:277]: Yeah - we should move this to utils and add a few more direct (not interaction) tests with different ID numbers to ensure we get the right result.

File: go-app-personal.js [hunk 2]
```diff
@@ -108,8 +116,13 @@ go.app = function() {
                 ],
 
                 next: function(choice) {
+                    self.contact.extra.language_choice = choice.value;
+
                     return self.im.user.set_lang(choice.value)
                     .then(function() {
+                        return self.im.contacts.save(self.contact);
+                    })
+                    .then(function() {
```
Comment (smn) [inline go-app-personal.js]: let's cut a separate PR for future refactorings where we fail fast if `.save(contact)` fails. Currently we're assuming that'll always be ok.

File: go-app-personal.js [hunk 3]
```diff
@@ -128,10 +141,15 @@ go.app = function() {
...
                 next: function(choice) {
-                    return {
-                        yes: 'states:id_type',
-                        no: 'states:end_not_pregnant'
-                    } [choice.value];
+                    self.contact.extra.suspect_pregnancy = choice.value;
+
+                    return self.im.contacts.save(self.contact)
+                    .then(function() {
+                        return {
```
Comment (smn) [inline go-app-personal.js]: nit pick but the lack of indentation had me confused.

File: go-app-personal.js [hunk 4]
```diff
@@ -187,13 +210,43 @@ go.app = function() {
...
+                        self.contact.extra.birth_month = birth_month.slice(1,2);
+                    } else {
+                        self.contact.extra.birth_month = birth_month;
+                    }
+                    
+                    if (parseInt(birth_day) <= 9) {
+                        self.contact.extra.birth_day = birth_day.slice(1,2);
+                    } else {
+                        self.contact.extra.birth_day = birth_day;
+                    }
```
Comment (smn) [inline go-app-personal.js]: Wouldn't it be easier to turn the `birth_year`, `birth_month` and `birth_day` with [moment.js](http://momentjs.com/docs/#/parsing/string-format/)? It's available in the sandbox.

Comment (smn) [inline go-app-personal.js]:
``` javascript
> moment = require('moment');
> moment('121314', 'MMDDYY').isValid();
true
> moment('121314', 'MMDDYY').format();
'2014-12-13T00:00:00+02:00'
> moment('141312', 'MMDDYY').isValid();
false
> moment('141312', 'MMDDYY').format();
'Invalid date'
```

File: go-app-personal.js [hunk 5]
```diff
@@ -187,13 +210,43 @@ go.app = function() {
...
+
+                    self.contact.extra.dob = (self.contact.extra.birth_year +
+                        '-' + self.contact.extra.birth_month + 
+                        '-' + self.contact.extra.birth_day);
```
Comment (smn) [inline go-app-personal.js]: if using moment then `the_date.format('YYYY-MM-DD');`

File: go-app-personal.js [hunk 6]
```diff
@@ -290,13 +372,18 @@ go.app = function() {
...
-                        name: 'states:end_success',
-                        creator_opts: {
-                            retry: opts.retry
-                        }
-                    };
+                next: function(content) {
+                    self.contact.extra.birth_day = content;
+                    self.contact.extra.dob = (self.im.user.answers['states:birth_year'] + 
+                        '-' + self.im.user.answers['states:birth_month'] +
+                        '-' + content);
```
Comment (smn) [inline go-app-personal.js:424]: We probably can find a use for moment.js here too.

File: go-app-personal.js (hunk 2 above)
Comment (gsvr) [inline go-app-personal.js]: Created issue #24

**Comment by smn** [comment]
ready for review?

**Comment by imsickofmaps** [comment]
@smn I'm just about to do an internal review

File: go-app-chw.js [hunk 7]
```diff
@@ -100,6 +125,7 @@ go.app = function() {
     var ChoiceState = vumigo.states.ChoiceState;
     var EndState = vumigo.states.EndState;
     var FreeText = vumigo.states.FreeText;
+    // var Q = require('q');
```
Comment (imsickofmaps) [inline go-app-chw.js]: remove stray lib - will be added back when metrics done

File: src/personal.js [hunk 8]
```diff
@@ -246,13 +298,21 @@ go.app = function() {
...
+                next: function(content) {
+                    if (content.length === 1) {
+                        content = '0' + content;
+                    }
+                    self.contact.extra.birth_day = content;
+                    self.contact.extra.dob = (self.im.user.answers['states:birth_year'] + 
+                        '-' + self.im.user.answers['states:birth_month'] +
+                        '-' + content);
+
+                    return self.im.contacts.save(self.contact)
```
Comment (imsickofmaps) [inline src/personal.js:310]: Indent

File: src/personal.js [hunk 9]
```diff
@@ -246,13 +298,21 @@ go.app = function() {
...
+                    .then(function() {
```
Comment (imsickofmaps) [inline src/personal.js]: Also need to remove the `working_on` extra from the user

File: src/personal.js (hunk 8 above)
Comment (gsvr) [inline src/personal.js:310]: improved all over

**Comment by imsickofmaps** [comment]
@smn ready for review

File: go-app-chw.js [hunk 10]
```diff
@@ -115,7 +140,15 @@ go.app = function() {
             return self.im.contacts
                 .for_user()
                 .then(function(user_contact) {
-                   self.contact = user_contact;
+                    if ((user_contact.extra.working_on !== undefined) && (user_contact.extra.working_on !== "")){
```
Comment (smn) [inline go-app-chw.js]: You'll like [`_.isUndefined()`](http://devdocs.io/lodash/index#isUndefined) from lodash.

File: go-app-chw.js [hunk 11]
```diff
@@ -87,11 +87,36 @@ go.utils = {
...
+        var birth_day = id.slice(4,6);
+
+        // assume if born before 1950 they won't get pregnant
+        // assumption necessary due to sa id number duplication
+        if (parseInt(birth_year) >= 50) {
+           birth_year = '19' + birth_year;
+        } else {
+           birth_year = '20' + birth_year;
+        }
+        return birth_year + '-' + birth_month + '-' + birth_day;
```
Comment (smn) [inline go-app-chw.js]: rather:

``` javascript
return moment(id.slice(0,6), 'YYMMDD').format('YY-MM-DD');
```

Will raise an error if invalid.

File: go-app-chw.js [hunk 12]
```diff
@@ -87,11 +87,36 @@ go.utils = {
...
     },
 
+    normalise_sa_msisdn: function(msisdn) {
+        normalised_no = '0' + msisdn.slice(3,12);
+        return normalised_no;
+    },
+
+    denormalise_sa_msisdn: function(msisdn) {
+        denormalised_no = '+27' + msisdn.slice(1,10);
+        return denormalised_no;
```
Comment (smn) [inline go-app-chw.js:113]: We'd call `+27...` normalized, not denormalized.

File: go-app-chw.js [hunk 13]
```diff
@@ -115,7 +140,15 @@ go.app = function() {
             return self.im.contacts
                 .for_user()
                 .then(function(user_contact) {
-                   self.contact = user_contact;
+                    if ((user_contact.extra.working_on !== undefined) && (user_contact.extra.working_on !== "")){
+                        self.user = user_contact;
+                        return self.im.contacts.get(user_contact.extra.working_on, {create: true})
```
Comment (smn) [inline go-app-chw.js:143]: what's `self.user` when not working on another contact? Can we set that to a default somewhere?

File: go-app-chw.js (hunk 12 above)
Comment (smn) [inline go-app-chw.js:113]: Well, both are normalized they're just different forms, I'd suggest you rename it to something that `[readable|friendly|human]_msisdn()` or something?

File: go-app-chw.js [hunk 14]
```diff
@@ -185,13 +220,16 @@ go.app = function() {
...
-                next: function() {
-                    return {
-                        name: 'states:id_type',
-                        creator_opts: {
-                            retry: opts.retry
-                        }
-                    };
+                next: function(content) {
+                    msisdn = go.utils.denormalise_sa_msisdn(content);
+                    self.contact.extra.working_on = msisdn;
```
Comment (smn) [inline go-app-chw.js:224]: This is where the confusion around normalized / denormalized came from.

File: go-app-chw.js [hunk 15]
```diff
@@ -238,13 +281,21 @@ go.app = function() {
...
-                    };
+                next: function(content) {
+                    self.contact.extra.sa_id = content;
+
+                    var id_date_of_birth = go.utils.extract_id_dob(content);
+                    self.contact.extra.birth_year = id_date_of_birth.slice(0,4);
+                    self.contact.extra.birth_month = id_date_of_birth.slice(5,7);
+                    self.contact.extra.birth_day = id_date_of_birth.slice(8,10);
+                    self.contact.extra.dob = id_date_of_birth;
+
```
Comment (smn) [inline go-app-chw.js:291]:
``` javascript
moment('2014-12-12', 'YYYY-MM-DD').format('D');
```

File: go-app-chw.js [hunk 16]
```diff
@@ -293,18 +362,20 @@ go.app = function() {
                 question: question,
 
                 check: function(content) {
-                    if (!go.utils.check_number_in_range(content, 1900, go.utils.get_today(self.im.config.testing_date).getFullYear())) {
+                    if (!go.utils.check_number_in_range(content, 1900, go.utils.get_today(self.im.config.testing_today).getFullYear())) {
```
Comment (smn) [inline go-app-chw.js]: why reference `testing_today` in production code?

File: go-app-chw.js [hunk 17]
```diff
@@ -315,7 +386,16 @@ go.app = function() {
 
                 choices: go.utils.make_month_choices($, 0, 12),
 
-                next: 'states:birth_day'
+                next: function(choice) {
+                    self.contact.extra.birth_month = choice.value;
```
Comment (smn) [inline go-app-chw.js:389]: I'm assuming it's according to scope that we both ask and do-not ask for these explicitly?

File: go-app-chw.js [hunk 18]
```diff
@@ -342,13 +422,21 @@ go.app = function() {
...
-                        }
-                    };
+                next: function(content) {
+                    if (content.length === 1) {
+                        content = '0' + content;
+                    }
+                    self.contact.extra.birth_day = content;
+                    self.contact.extra.dob = (self.im.user.answers['states:birth_year'] + 
+                        '-' + self.im.user.answers['states:birth_month'] +
+                        '-' + content);
```
Comment (smn) [inline go-app-chw.js]: moment(...).format(...); please

File: go-app-chw.js [hunk 19]
```diff
@@ -367,7 +455,19 @@ go.app = function() {
...
                 next: function(choice) {
+                    self.contact.extra.language_choice = choice.value;
+
                     return self.im.user.set_lang(choice.value)
+                    // we may not have to run this for this flow
+                    .then(function() {
+                        return self.im.contacts.save(self.contact);
+                    })
+                    .then(function() {
+                        if (self.user !== undefined && self.user.extra.working_on !== undefined) {
```
Comment (smn) [inline go-app-chw.js]: `_.isUndefined()`

File: test/chw.test.js [hunk 20]
```diff
@@ -14,6 +14,8 @@ describe("app", function() {
             tester = new AppTester(app);
 
             tester
+                .setup.user.lang('en')
+                .setup.char_limit(180)
```
Comment (smn) [inline test/chw.test.js]: 180 seems to fail on MTN, please report when this happens so content can be updated to fit 160.

**Comment by smn** [comment]
done reviewing for now.

**Comment by smn** [comment]
@imsickofmaps for TZ check http://momentjs.com/docs/#/manipulating/timezone-offset/

File: go-app-chw.js (hunk 17 above)
Comment (gsvr) [inline go-app-chw.js:389]: Yes, when the user selects their id type we either:
1. Ask for their dob directly if they select a None type ID
2. Extract their dob from ID if they have an SA ID
3. Don't collect their dob if they input a foreign Passport.
This as shown in user flow 7b.

Comment (smn) [inline go-app-chw.js:389]: thanks

**Comment by smn** [comment]
When can I expect to do the final review?

**Comment by imsickofmaps** [comment]
@smn ready

**Comment by smn** [comment]
:+1:

**Pull request merged**